File: jarvis/backend/app/core/tools/scheduler.py
"""Reminder scheduling: create_reminder, list_reminders, cancel_reminder,
plus an async `run_scheduler(on_fire)` ticker that fires due reminders.

Reminders persist to SQLite at `settings.tasks_db_path` so they survive a
process restart. All DB access goes through a tiny helper module-level
connection guarded by a lock since sqlite3 connections aren't safely shared
across concurrent awaits without one.
"""

from __future__ import annotations

import asyncio
import logging
import sqlite3
import time
import uuid
from collections.abc import Awaitable, Callable
from typing import Any

from app.config import settings
from app.core.tools.base import register
from app.models.schemas import ToolResult

logger = logging.getLogger("jarvis.tools.scheduler")

_TICK_INTERVAL_S = 5.0
_db_lock = asyncio.Lock()


def _connect() -> sqlite3.Connection:
    settings.tasks_db_path.parent.mkdir(parents=True, exist_ok=True)
    conn = sqlite3.connect(settings.tasks_db_path)
    conn.execute(
        """
        CREATE TABLE IF NOT EXISTS reminders (
            id TEXT PRIMARY KEY,
            text TEXT NOT NULL,
            due_at REAL NOT NULL,
            created_at REAL NOT NULL,
            fired INTEGER NOT NULL DEFAULT 0,
            cancelled INTEGER NOT NULL DEFAULT 0
        )
        """
    )
    conn.commit()
    return conn


async def _run_db(fn: Callable[[sqlite3.Connection], Any]) -> Any:
    """Run a synchronous sqlite operation off the event loop, serialized."""
    async with _db_lock:
        conn = _connect()
        try:
            return await asyncio.to_thread(_with_conn, conn, fn)
        finally:
            conn.close()


def _with_conn(conn: sqlite3.Connection, fn: Callable[[sqlite3.Connection], Any]) -> Any:
    result = fn(conn)
    conn.commit()
    return result


@register
class CreateReminderTool:
    name = "create_reminder"
    description = "Schedule a reminder to fire after a given number of seconds from now."
    schema: dict[str, Any] = {
        "type": "object",
        "properties": {
            "text": {"type": "string", "minLength": 1, "description": "Reminder text."},
            "delay_seconds": {
                "type": "number",
                "minimum": 1,
                "description": "How many seconds from now the reminder should fire.",
            },
        },
        "required": ["text", "delay_seconds"],
        "additionalProperties": False,
    }

    async def run(self, text: str, delay_seconds: float) -> ToolResult:
        reminder_id = str(uuid.uuid4())
        now = time.time()
        due_at = now + float(delay_seconds)

        def _insert(conn: sqlite3.Connection) -> None:
            conn.execute(
                "INSERT INTO reminders (id, text, due_at, created_at, fired, cancelled) "
                "VALUES (?, ?, ?, ?, 0, 0)",
                (reminder_id, text, due_at, now),
            )

        await _run_db(_insert)
        return ToolResult(
            ok=True,
            output=reminder_id,
            summary=f"Reminder {reminder_id} scheduled in {delay_seconds:.0f}s",
            meta={"id": reminder_id, "due_at": due_at},
        )


@register
class ListRemindersTool:
    name = "list_reminders"
    description = "List pending (not yet fired or cancelled) reminders."
    schema: dict[str, Any] = {
        "type": "object",
        "properties": {},
        "required": [],
        "additionalProperties": False,
    }

    async def run(self) -> ToolResult:
        def _select(conn: sqlite3.Connection) -> list[tuple[str, str, float]]:
            cur = conn.execute(
                "SELECT id, text, due_at FROM reminders WHERE fired = 0 AND cancelled = 0 ORDER BY due_at ASC"
            )
            return cur.fetchall()

        rows = await _run_db(_select)
        lines = [f"{rid}: {text!r} due at {due_at:.0f}" for rid, text, due_at in rows]
        return ToolResult(
            ok=True,
            output="\n".join(lines),
            summary=f"{len(rows)} pending reminder(s)",
            meta={"count": len(rows)},
        )


@register
class CancelReminderTool:
    name = "cancel_reminder"
    description = "Cancel a pending reminder by id."
    schema: dict[str, Any] = {
        "type": "object",
        "properties": {"id": {"type": "string", "minLength": 1}},
        "required": ["id"],
        "additionalProperties": False,
    }

    async def run(self, id: str) -> ToolResult:
        def _cancel(conn: sqlite3.Connection) -> int:
            cur = conn.execute(
                "UPDATE reminders SET cancelled = 1 WHERE id = ? AND fired = 0 AND cancelled = 0", (id,)
            )
            return cur.rowcount

        rowcount = await _run_db(_cancel)
        if rowcount == 0:
            return ToolResult(ok=False, output="", summary=f"No pending reminder with id {id!r}")
        return ToolResult(ok=True, output="", summary=f"Cancelled reminder {id!r}")


async def run_scheduler(
    on_fire: Callable[[str, str], Awaitable[None]],
    *,
    interval_s: float = _TICK_INTERVAL_S,
) -> None:
    """Poll SQLite forever, calling `on_fire(id, text)` for each due reminder.

    Intended to run as a long-lived background task started at app startup.
    Never raises out of the loop — a single tick's failure is logged and the
    ticker keeps going so one bad reminder can't kill scheduling entirely.
    """
    while True:
        try:
            due = await _pop_due_reminders()
            for reminder_id, text in due:
                try:
                    await on_fire(reminder_id, text)
                except Exception:  # noqa: BLE001 - one bad callback must not kill the ticker
                    logger.exception("on_fire callback failed for reminder %s", reminder_id)
        except Exception:  # noqa: BLE001 - scheduler loop must never die
            logger.exception("scheduler tick failed")
        await asyncio.sleep(interval_s)


async def _pop_due_reminders() -> list[tuple[str, str]]:
    now = time.time()

    def _select_and_mark(conn: sqlite3.Connection) -> list[tuple[str, str]]:
        cur = conn.execute(
            "SELECT id, text FROM reminders WHERE fired = 0 AND cancelled = 0 AND due_at <= ?", (now,)
        )
        rows = cur.fetchall()
        if rows:
            ids = [r[0] for r in rows]
            conn.executemany("UPDATE reminders SET fired = 1 WHERE id = ?", [(i,) for i in ids])
        return rows

    return await _run_db(_select_and_mark)

File: jarvis/server/src/tools/scheduler.ts
/**
 * Reminder scheduling: create_reminder, list_reminders, cancel_reminder,
 * plus `runScheduler(onFire)` — a long-lived ticker that fires due
 * reminders.
 *
 * Reminders persist to SQLite at `config.tasksDbPath` so they survive a
 * process restart. `better-sqlite3` is synchronous, so no locking helper is
 * needed the way the Python edition needs one around its async sqlite3
 * connection.
 */

import Database from 'better-sqlite3';
import { randomUUID } from 'node:crypto';
import { promises as fs } from 'node:fs';
import path from 'node:path';

import { config } from '../config.js';
import { registry } from './base.js';
import type { Tool, ToolResult } from '../types.js';

const TICK_INTERVAL_MS = 5_000;

interface ReminderRow {
  id: string;
  text: string;
  due_at: number;
  created_at: number;
  fired: number;
  cancelled: number;
}

let db: Database.Database | null = null;

function getDb(): Database.Database {
  if (db) return db;
  db = new Database(config.tasksDbPath);
  db.pragma('journal_mode = WAL');
  db.exec(`
    CREATE TABLE IF NOT EXISTS reminders (
      id TEXT PRIMARY KEY,
      text TEXT NOT NULL,
      due_at REAL NOT NULL,
      created_at REAL NOT NULL,
      fired INTEGER NOT NULL DEFAULT 0,
      cancelled INTEGER NOT NULL DEFAULT 0
    )
  `);
  return db;
}

async function ensureDbDir(): Promise<void> {
  await fs.mkdir(path.dirname(config.tasksDbPath), { recursive: true });
}

function fail(summary: string): ToolResult {
  return { ok: false, output: '', summary };
}

const createReminderTool: Tool = {
  name: 'create_reminder',
  description: 'Schedule a reminder to fire after a given number of seconds from now.',
  parameters: {
    type: 'object',
    properties: {
      text: { type: 'string', minLength: 1, description: 'Reminder text.' },
      delay_seconds: {
        type: 'number',
        minimum: 1,
        description: 'How many seconds from now the reminder should fire.',
      },
    },
    required: ['text', 'delay_seconds'],
    additionalProperties: false,
  },
  async run(args): Promise<ToolResult> {
    const text = args.text as string;
    const delaySeconds = args.delay_seconds as number;

    await ensureDbDir();
    const id = randomUUID();
    const now = Date.now() / 1000;
    const dueAt = now + delaySeconds;

    getDb()
      .prepare(
        'INSERT INTO reminders (id, text, due_at, created_at, fired, cancelled) VALUES (?, ?, ?, ?, 0, 0)',
      )
      .run(id, text, dueAt, now);

    return {
      ok: true,
      output: id,
      summary: `Reminder ${id} scheduled in ${Math.round(delaySeconds)}s`,
      meta: { id, due_at: dueAt },
    };
  },
};

const listRemindersTool: Tool = {
  name: 'list_reminders',
  description: 'List pending (not yet fired or cancelled) reminders.',
  parameters: { type: 'object', properties: {}, required: [], additionalProperties: false },
  async run(): Promise<ToolResult> {
    await ensureDbDir();
    const rows = getDb()
      .prepare('SELECT id, text, due_at FROM reminders WHERE fired = 0 AND cancelled = 0 ORDER BY due_at ASC')
      .all() as Array<Pick<ReminderRow, 'id' | 'text' | 'due_at'>>;

    const lines = rows.map((r) => `${r.id}: ${JSON.stringify(r.text)} due at ${Math.round(r.due_at)}`);
    return {
      ok: true,
      output: lines.join('\n'),
      summary: `${rows.length} pending reminder(s)`,
      meta: { count: rows.length },
    };
  },
};

const cancelReminderTool: Tool = {
  name: 'cancel_reminder',
  description: 'Cancel a pending reminder by id.',
  parameters: {
    type: 'object',
    properties: { id: { type: 'string', minLength: 1 } },
    required: ['id'],
    additionalProperties: false,
  },
  async run(args): Promise<ToolResult> {
    const id = args.id as string;
    await ensureDbDir();
    const info = getDb()
      .prepare('UPDATE reminders SET cancelled = 1 WHERE id = ? AND fired = 0 AND cancelled = 0')
      .run(id);

    if (info.changes === 0) {
      return fail(`No pending reminder with id ${JSON.stringify(id)}`);
    }
    return { ok: true, output: '', summary: `Cancelled reminder ${JSON.stringify(id)}` };
  },
};

registry.register(createReminderTool);
registry.register(listRemindersTool);
registry.register(cancelReminderTool);

/**
 * Poll SQLite forever, calling `onFire(id, text)` for each due reminder.
 *
 * Intended to run as a long-lived background task started at server
 * startup. Never throws out of the loop — a single tick's failure is
 * logged and the ticker keeps going so one bad reminder can't kill
 * scheduling entirely. Returns a stop function.
 */
export function runScheduler(
  onFire: (id: string, text: string) => Promise<void>,
  intervalMs: number = TICK_INTERVAL_MS,
): () => void {
  let stopped = false;

  const tick = async (): Promise<void> => {
    if (stopped) return;
    try {
      await ensureDbDir();
      const due = popDueReminders();
      for (const { id, text } of due) {
        try {
          await onFire(id, text);
        } catch (exc) {
          console.error(`[scheduler] onFire callback failed for reminder ${id}:`, exc);
        }
      }
    } catch (exc) {
      console.error('[scheduler] tick failed:', exc);
    }
  };

  const timer = setInterval(() => {
    void tick();
  }, intervalMs);
  // Don't hold the process open just for the scheduler in short-lived contexts.
  timer.unref?.();

  return () => {
    stopped = true;
    clearInterval(timer);
  };
}

function popDueReminders(): Array<{ id: string; text: string }> {
  const now = Date.now() / 1000;
  const database = getDb();
  const rows = database
    .prepare('SELECT id, text FROM reminders WHERE fired = 0 AND cancelled = 0 AND due_at <= ?')
    .all(now) as Array<{ id: string; text: string }>;

  if (rows.length > 0) {
    const markFired = database.prepare('UPDATE reminders SET fired = 1 WHERE id = ?');
    const markAll = database.transaction((ids: string[]) => {
      for (const id of ids) markFired.run(id);
    });
    markAll(rows.map((r) => r.id));
  }
  return rows;
}
